### DISCUSSION TITLE
Too many open file issue

### DISCUSSION THREAD

**Discussion by L3AZH** [relayrides/pushy#1062] [Q&A] [answered]
From the version v.0.14.0 
I have a question about the APNS client as the following:
If I create a APNS client as the following code 

```
apnsClient = apnsClientBuilder.setProxyHandlerFactory(proxyHandlerFactory).
                setApnsServer("api.server.....").
                setSigningKey(<some credential>).
                build();
```
And I using the sync mode ( wait for the APNS server response and the timeout is about 7s)
1. So the default thread number from NioEventLoopGroup will be 1 and capacity of Channel pool will be 1 right ?
2. From the 1 above if a large request were created to send notification to APNS server. Then for each request, a new channel will be created first then other request will go into pending queue ( because the channel capacity is 1) until the first request is finished and the channel is released to the channel poll is that correct ?

I also made a load test and  monitored the eventpoll number from the application by `lsof -p <pid> | grep eventpoll | wc -l` and saw that the number of eventpoll were increased during the time of the load test. Did the config above affected or related to this ?

**Comment by jchambers** [comment]
I'm not entirely sure I understand the scenario you're describing; can you please clarify what you mean by "a large request?"

**Comment by L3AZH** [comment]
Let me be more specific,
I have an application that handles the sending notifications to the APNS server which is using Pushy v0.14.0(FYI the application  deployed in the JBoss server on a Linux VM)
The APNS client config is the following code:
```
apnsClient = apnsClientBuilder.setProxyHandlerFactory(proxyHandlerFactory).
                setApnsServer("api.server.....").
                setSigningKey(<some credential>).
                build();

```
I use a sync mode when sending the requests and receive the responses from the ASPN server like the below:
```
final CompletableFuture sendNotificationFuture = apnsClient.sendNotification(pushNotification);
final PushNotificationResponse response = sendNotificationFuture.get(timoue = 7000,TimeUnit = MILISEC);
```

When I performed the load test ( maybe about 87500 sample requests to an API that handles the sending  notifications) there are things that I noticed:
I have checked the open file descriptor of the application by cmd : 
`lsof -p <pid> | grep eventpoll | wc -l`
Before the test, the number of eventpolls was about 73 or 75 
During the test, the number kept increasing by about 1 or 2 each
By the end of the test, the number ended up at 220
And I performed the test many times.
In some cases, after 40 minutes or 1 hour, the number of eventpoll will decrease to the original number( 73-75 same as the number before the test)
But in some cases, the number did not decrease at all, I had to restart the JBoss server and deploy the application again 
FYI: Due to the restriction of resources on the Linux VM, the setting for max open file in the system is 5000 ( check with cmd: `ulimit -a`)
So I have some concerns here:
Does the APNS client create a new open file descriptor for each sending notification job ( using "Creating a new open file descriptor" may confuse, so I think that an open file descriptor can be understood as a channel or event loop, I know the channel and event loop are a different concept in Netty)
Because I'm using the sync method so does it improve when I switch to async? As I understand when reading your code, the default Thread in NioEventLoopGroup, Capacity of channels, and concurrent connection number when init an APNS client will be 1, so when using async it may increase the release channel speed, is that correct?

**Reply by jchambers** [reply]
I'm still not entirely sure I understand, but it sounds like you might be affected by #853, which was fixed a long time ago. Please update to the latest version of Pushy, and please let us know if the problem persists after updating.

**Reply by L3AZH** [reply, marked as answer]
It turns out the event polls number is increased due to another service, I've isolated the service that uses APNS pushy to another VM, and the event polls number is quite ok on the new VM, not increasing
